feat: Closed

ryangamerdev opened the pull request:
Removed.

ariane-emory commented:
Not really in favour of changing the defaults, new behaviour ought be opt-in IMO.

ryangamerdev committed:
feat: add collapse compaction mode

Adds a new 'collapse' compaction mode that preserves recent context while
summarizing older messages. Unlike standard compaction which summarizes the
entire conversation, collapse extracts the oldest 65% of tokens, summarizes
them at a breakpoint, and leaves the newest 35% untouched.

Changes:
- Add processCollapse() for breakpoint-based summarization
- Add Identifier.createLike() for inserting messages at past timestamps
- Add detectFormat() for backward compatibility with existing sessions
- Add configurable ratios: trigger (0.85), extractRatio (0.65), recentRatio (0.15)
- Add TUI toggle for switching between standard/collapse modes
- Fix isOverflow() to include cache.write tokens
- Fix link.tsx underline prop error from #6317

Default remains 'standard' - users can opt-in to collapse via config or TUI toggle.

ryangamerdev commented:
> Not really in favour of changing the defaults, new behaviour ought be opt-in IMO.

Fixed it. Thank you for the feedback.

ryanwyler commented:
**Before Compaction: 170,424 Tokens**
**After Compaction: 67,289 Tokens** 

### **60% Reduction** in token consumption and **100% MEMORY RETENTION**

<img width="1825" height="1010" alt="before-collapse-compact" src="https://github.com/user-attachments/assets/55d9de61-db26-4af0-94a0-581db8a5240d" />

### After collapse compact, AI Agent able to produce pages and pages of details about the project on many topics without the use of any tools and no reading of documents.

<img width="1825" height="1010" alt="after-collapse-screen1" src="https://github.com/user-attachments/assets/0020f72f-603e-483b-904c-2520e57fe5e9" />
<img width="1825" height="1010" alt="after-collapse-screen2" src="https://github.com/user-attachments/assets/0108a812-1a14-41ef-82e8-85f3caa10f49" />
<img width="1825" height="1010" alt="after-collapse-screen3" src="https://github.com/user-attachments/assets/849d1493-1cef-4ee2-a8e6-74f2a752e662" />
<img width="1825" height="1010" alt="after-collapse-screen4" src="https://github.com/user-attachments/assets/12020cf4-6633-44e1-9159-165273b2c05c" />
<img width="1825" height="1010" alt="after-collapse-screen5" src="https://github.com/user-attachments/assets/6c0bc6f7-1083-4596-8e5a-1a5fe21f8196" />

ryangamerdev committed:
fix: resolve broken forked sessions with compactions due to missing parent-child message references

When forking a session, message IDs are regenerated but parentID references in assistant messages were not updated to point to the new IDs. This broke features that rely on parent-child relationships, such as compaction breakpoints.

## Problem

Forking a compacted session would fail with:
```
prompt is too long: 203573 tokens > 200000 maximum
```

The forked session didn't honor the compaction breakpoint because the parent-child message relationships were broken.

## Fix

Added an ID mapping that tracks old ID -> new ID, then updates parentID references when cloning assistant messages.

## Changes

- packages/opencode/src/session/index.ts - Add ID mapping in fork function to preserve parent-child relationships

ryangamerdev committed:
feat: include previous summaries in collapse compaction

Adds support for merging historical context from previous compaction summaries
into new collapse summaries. This prevents loss of important information across
multiple compaction cycles.

New config options:
- summaryMaxTokens: target token count for summary output (default: 10000)
- previousSummaries: number of previous summaries to include (default: 3)

Changes:
- Add getPreviousSummaries() to fetch historical summaries from unfiltered messages
- Add extractSummaryText() to extract text content from summary parts
- Update collapse prompt to include previous summaries section
- Add merge instructions to preserve historical context

ryangamerdev committed:
feat: add token budget for previous summaries and improve prompt structure

- Calculate available token budget before fetching previous summaries
- Use XML-style delimiters for clearer prompt structure
- Add debug logging with [BUDGET_CALC] tag for troubleshooting
- Prevents context overflow when including previous summaries
